### PR TITLE
[AIRFLOW-5172] Add choice of interval edge scheduling

### PR DESCRIPTION
**Pull request by iroddis**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5172

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This change introduces the attribute `schedule_interval_edge`, a string containing either 'start' or 'end', to DAGs. The scheduler uses the value to determining if a DAG should be scheduled at the start or the end of the schedule interval.

A parameter with the same name was also added to the default_airflow.cfg in the [scheduler] section. 

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

tests.jobs.test_scheduler_job:SchedulerJobTest.test_schedule_interval_edge

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

### CODE REVIEW

**Comment by ashb** [comment]
This needs some docs re-writing as a result -https://airflow.apache.org/scheduler.html (which lives under docs/ somewhere) at the least.

**Comment by ashb** [comment]
See also https://github.com/apache/airflow/pull/4768/files for previous discussion about this. I think I'm okay with this generally, but we'll need to make sure the docs about it are clear - i.e. more than just a comment in a config file; likely a section in that scheduler page with some images/pictures of timelines.

Page 14 of https://drive.google.com/open?id=1DVN4HXtOC-HXvv00sEkoB90mxLDnCIKc may be a starting point for that image.

**Comment by iroddis** [comment]
Yup, I'll make those updates. After more thought, I'm also going to change the variable from a str to a bool, and likely the name to something like schedule_at_interval_end. Any suggestions for better names?

File: docs/scheduler.rst
Comment (ashb) [inline docs/scheduler.rst]:
```suggestion
  2019-08-16 17:00 will start running just after 2019-08-16 17:59:59 -- i.e. once that hour is over.
```

File: docs/scheduler.rst
Comment (ashb) [inline docs/scheduler.rst]: This is an unfortunate abbreviation in British English :)

```suggestion
    start_of_day = DAG(
```

File: airflow/config_templates/default_airflow.cfg
Comment (ashb) [inline airflow/config_templates/default_airflow.cfg]:
```suggestion
# Should DAGs be scheduled to execute at the 'start' or 'end'
```

File: airflow/config_templates/default_airflow.cfg
Comment (ashb) [inline airflow/config_templates/default_airflow.cfg]: I wonder if `schedule_at_interval_end` is a better/clearer name.

I'm probably splitting hairs

Comment (iroddis) [reply airflow/config_templates/default_airflow.cfg]: I'm all for hair splitting if it means clearer usage down the road, and is more consistent with established nomenclature.

Part of my confusion is that `schedule` seams to have multiple meanings: A job scheduled daily will be scheduled to run once the day has finished. So there's the declared schedule, and then the implied schedule (execution time).

I have no vested interest in the name, just let me know if you'd like to change it.

Either way, maybe it would be worth adding (in a separate PR) a dictionary of terms and precise definitions? I'd be happy submit that after this.

Comment (ashb) [reply airflow/config_templates/default_airflow.cfg]: All these terms are confusing and overlapping.

The reason I wanted to avoid `execute` is because the `execution_date` property of the task  will remain constant non matter what this setting. So I think this does need to change, but I'm not at all convinced my wording is better.

> Either way, maybe it would be worth adding (in a separate PR) a dictionary of terms and precise definitions? I'd be happy submit that after this.

Yes, that would be great!

Comment (iroddis) [reply airflow/config_templates/default_airflow.cfg]: I've updated the name, I like it, and it makes sense. Also made a few more very small documentation changes, and squashed all the commits. Please let me know if you'd like to change anything else. 

Do you think there's any way this could get cherry picked for the next stable release? Right now I'm using a hacked up local version to get this behaviour, and it would be great to get back to upstream stable.

**Comment by iroddis** [comment]
I'm not sure why CI failed on this push ... is it possible to re-run them?

-----
edit: Nevermind. I see the same tests failing in my local repo. They also fail when I test against my current master branch.

**Review by ashb — APPROVED** [review]
Happy with the code on this change, I'm opening a thread on the mailing list to discuss this.

(Committers: please don't merge without checking that thread first)

**Comment by aoen** [comment]
Really like the spirit of this change. My vote would be to change the flag to schedule_at_interval_start and default it to false for now as scheduling_at_interval_start is the more intuitive behavior. What are people's feelings on changing the default execution to schedule interval start and communicating this to existing users in the Updating notes so that they can preserve the old behavior? Could potentially cause headaches for users who don't read the notes but I think it might make sense to bite the bullet at some point for more intuitive behavior overall for new users.

I haven't looked that carefully, but I think there are potentially a lot more places in the code that need to be changed that use dag.following_schedule(), e.g. latest_only_operator, cli.py. In fact it might make more sense to make the change in the following_schedule function itself. There are also functions previous_schedule that might need to be modified too.

File: airflow/config_templates/default_airflow.cfg
Comment (feng-tao) [inline airflow/config_templates/default_airflow.cfg:528]: +1 on the default to be false as this is the intuitive / common understanding on how Airflow scheduling works.

**Review by feng-tao — CHANGES_REQUESTED** [review]

File: airflow/config_templates/default_airflow.cfg
Comment (iroddis) [reply airflow/config_templates/default_airflow.cfg:528]: The PR is designed to have zero impact to current behaviours, to avoid surprises for devops everywhere and to increase the chance that the PR could be accepted for the next 1.x.x stable release.

Would it make more sense to leave this PR as is, and then submit a second PR to make the default False at a later date, for 2.0?

File: airflow/jobs/scheduler_job.py
Comment (JCoder01) [inline airflow/jobs/scheduler_job.py]: Does similar logic need to be applied to manage_slas?

**Comment by ashb** [comment]
> Really like the spirit of this change. My vote would be to change the flag to schedule_at_interval_start and default it to false for now as scheduling_at_interval_start is the more intuitive behavior. What are people's feelings on changing the default execution to schedule interval start and communicating this to existing users in the Updating notes so that they can preserve the old behavior? Could potentially cause headaches for users who don't read the notes but I think it might make sense to bite the bullet at some point for more intuitive behavior overall for new users.

It might be more intuitive to new users but this would mean that every single dag in existence right now would need to be updated, so it's a very disrputive change.

 
> I haven't looked that carefully, but I think there are potentially a lot more places in the code that need to be changed that use dag.following_schedule(), e.g. latest_only_operator, cli.py. In fact it might make more sense to make the change in the following_schedule function itself. There are also functions previous_schedule that might need to be modified too.

Yes, that seems sensible. @iroddis Could you take a look at this and see if it makes sense or other places need updating too?

**Comment by dstandish** [comment]
@iroddis instead why not just add a `run_date` property to DagRun, and make it available in template context?
It seems to me this would provide the clarity you seek without having to add additional configuration or changes to scheduler.   
Adding run date would make explicit the run date vs execution date distinction.  
Adding run date would allow execution_date to be consistently defined (as opposed to sometimes meaning run date and other times meaning last-run-date).
Then in the same task one task might use "execution_date" while for another task "run_date" might be more appropriate.

**Comment by iroddis** [comment]
Sorry for the random commits on this. Getting the unit tests running is still a challenge, especially on osx where I'm doing these changes currently.

**Comment by iroddis** [comment]
The assumption that the run time for an interval is the dag.following_schedule(dttm) is baked in all over the place. I've found some more cases, and rebased off of current master.

Making the change in following_schedule() was not the right route; that method is frequently used to iterate through run times. I instead added a dag.period_end(dttm) method to use that incorporates the new `schedule_at_interval_end` parameter.

As of this last checkin, all tests seem to be passing except some that look like race conditions, as they'll pass and not pass from run to run.

**Comment by vardancse** [comment]
Pretty much interested in this functionality. Any upcoming plans to make this changes part of release before v2.0, if not we'd like to cherry pick this commit for v1.10.5 and see if it works. BTW do we have any guidance, if we can pick this up and merge with older version specifically v1.10.2?

**Comment by iroddis** [comment]
I haven't run the tests or tried to backport the functionality on older releases. All of the tests are currently passing, but I'm sure there are edge cases. I'd love to see this in a release, maybe tagged as an alpha feature, to get some wider testing.

File: airflow/jobs/scheduler_job.py
Comment (dstandish) [inline airflow/jobs/scheduler_job.py:651]: it seems there is an inconsistency in the language here.  

The function is called `period_end`, yet sometimes it returns the start of the interval. 

The name `period_end` for this variable reflects an assumption that dags are scheduled after the _end_ of the interval.  

That's why the code checks that `period_end <= timezone.utcnow()`: end of interval is in the past.

But language of this PR is in conflict with that.  

The parameter `schedule_at_interval_end` implies that the _interval_ doesn't change, but _where we schedule_ does.  So, we may schedule at start or end of "the interval".  But as written, if `schedule_at_interval_end=False`, it will in general be the case that `period_end==excecution_date`, which implies that exec date is the end of "the interval" and not the start, and this is a contradiction.

It seems that what `period_end` represents in this code is more like `run_after_dttm` -- the datetime before which the dag may not be scheduled.  When `schedule_at_interval_end` is True, we can run after exec date + 1 interval; otherwise, we can run after exec date.  

So in this bit of code, we probably don't even need `period_end()` because we could do this:
```
            run_after_dttm = None
            if dag.schedule_interval == '@once':
                run_after_dttm = next_run_date
            elif next_run_date and not self.schedule_at_interval_end:
                run_after_dttm = next_run_date
            elif next_run_date and self.schedule_at_interval_end:
                run_after_dttm = dag.following_schedule(next_run_date)
```
And this:
```
            if next_run_date and run_after_dttm and run_after_dttm <= timezone.utcnow():
```

But elsewhere, it seems that `period_end()` function is used to mean `min_run_date` or `target_run_date` or `run_after_date`.  Perhaps a name like this would be clearer.

**Review by dstandish — CHANGES_REQUESTED** [review]

File: airflow/jobs/scheduler_job.py
Comment (iroddis) [reply airflow/jobs/scheduler_job.py:651]: Nomenclature has absolutely been a pain in this change. The entire vocabulary around scheduling seems to have exploded, and I can't find a clear definition for all the terms. This is well outside the scope of this change, but having well defined terms for things like:

period_start_ts - Timestamp of the start of the period to be processed
period_end_ts - Timestamp of the end of the period to be processed
scheduled_execution_ts - Timestamp of when the execution should start
execution_ts - Timestamp of actual execution start

Many of the current names, like start_date, are datetimes, not dates. It's also not immediately obvious what "start" means: is it period_start_ts, or execution_ts?

I'm happy to change the functions to be whatever is popular, but I think there'd be real value in defining, clearly, and in a single spot, all of the timestamps around a particular DAG's execution, giving them meaningful names, and consolidating on that.

Comment (dstandish) [reply airflow/jobs/scheduler_job.py:651]: Yeah i'm not saying you are responsible for all confusing naming, just suggesting not to add to it ;)

`period_end` is a new function. 

And I think it is definitely in scope to change the local `period_end` variables since you are, in effect, changing their meaning in this PR.

**Comment by aoen** [comment]
> > Really like the spirit of this change. My vote would be to change the flag to schedule_at_interval_start and default it to false for now as scheduling_at_interval_start is the more intuitive behavior. What are people's feelings on changing the default execution to schedule interval start and communicating this to existing users in the Updating notes so that they can preserve the old behavior? Could potentially cause headaches for users who don't read the notes but I think it might make sense to bite the bullet at some point for more intuitive behavior overall for new users.
> 
> It might be more intuitive to new users but this would mean that every single dag in existence right now would need to be updated, so it's a very disrputive change.

@ashb To clarify, the idea here is to set a cluster-level default in the airflow.cfg to use the new semantics (execution date = start date), then anyone who owns an Airflow cluster with many customers can just make sure to change the default value back to the old semantics when we change the cluster-level default; no dags would need to be updated on such clusters. Anybody setting up Airflow for the first time would have sensical semantics, and maybe we can even deprecate the old semantics after some point in time since I don't think many new users will use it.

**Comment by ashb** [comment]
@aoen That sounds sensible, so long as we clearly "sign-post" the change for users when upgrading somehow. The default could only change from 2.0 though.

Are we still talking about having a per-dag setting to, or are you thinking that we just have a single global config setting?

**Comment by ryw** [comment]
do we want to include this in 2.0?

**Comment by h1-nik** [comment]
@mistercrunch @kaxil @potiuk @mik-laj @turbaszek @ashb @iroddis @ryw 
This is a much needed feature! The discrepency between triggered jobs + scheduled jobs not being able to line up on execution date causes a lot of unexpected scheduling issues.

**Comment by freddy-criollo** [comment]
@mistercrunch @kaxil @potiuk @mik-laj @turbaszek @ashb @iroddis @ryw
This is feature would be great!

**Comment by ryw** [comment]
@iroddis I'm pushing to get this into 2.0.  

Do you have time to iterate on the PR over the next week? Addressing @dstandish concerns, etc, rebasing it on the HA scheduler code (which we will merge very soon). 

If you don't have the time, we can take the PR over to fix it up and get it merged.

**Comment by iroddis** [comment]
@ryw I’m sorry, I stopped working on this a long time ago, and I don’t know when I’ll have the time to pick it back up.  I think some of the comments were correct about subtler edge cases that weren’t accounted for with scheduling, but got lost in the weeds before I could figure them out.

**Pull request closed without merging**
